### ISSUE TITLE
Lost token if no connection (Cordova)

### ISSUE THREAD

**Issue by dflourusso** [lynndylanhurley/ng-token-auth#335]
I'm already logged in and have the credentials stored in the localStorage, but if i open the application without Internet connection it lost the stored credentials and the user must do login again.
It should not reset, because when the connection is available, i do not want to type user and password again.

**Comment by scottward** [comment]
I have the exact same issue. I'm using Ionic/Angular/Cordova. Any web app would be unlikely to see this because if they can load the web app, they can probably reach the server. However, mobile apps will often be loaded when there is no network.

As far as I can see, the issue has to do with how things are handled in the `error` section of the  `validateToken` function:
```
validateToken: (opts={}) ->
  unless @tokenHasExpired()
    $http.get(@apiUrl(opts.config) + @getConfig(opts.config).tokenValidationPath)
      .success((resp) =>
        authData = @getConfig(opts.config).handleTokenValidationResponse(resp)
        @handleValidAuth(authData)

        # broadcast event for first time login
        if @firstTimeLogin
          $rootScope.$broadcast('auth:email-confirmation-success', @user)

        if @oauthRegistration
          $rootScope.$broadcast('auth:oauth-registration', @user)

        if @mustResetPassword
          $rootScope.$broadcast('auth:password-reset-confirm-success', @user)

        $rootScope.$broadcast('auth:validation-success', @user)
      )
      .error((data) =>
        # broadcast event for first time login failure
        if @firstTimeLogin
          $rootScope.$broadcast('auth:email-confirmation-error', data)

        if @mustResetPassword
          $rootScope.$broadcast('auth:password-reset-confirm-error', data)

        $rootScope.$broadcast('auth:validation-error', data)

        @rejectDfd({
          reason: 'unauthorized'
          errors: if data? then data.errors else ['Unspecified error']
        })
      )
```
At the end, `rejectDfd` is called which in turn calls `invalidateTokens`.

I think what needs to happen is that we need to distinguish between "validation errors" and "network errors". One thought would be make it so if `data.errors` doesn't exist then we have a network or server error on our hands, not a validation error.

In that case, perhaps the call to `rejectDfd` could include an option like `doNotInvalidateToken: true`.

Not sure if this is the right approach since I'm not overly familiar with the codebase.

Comments?

**Comment by dflourusso** [comment]
I created a pull request https://github.com/lynndylanhurley/ng-token-auth/pull/336
Lets wait the merge now.

**Comment by Charlie-Hua** [comment]
@scottward I am also using Ionic and my workaround is to wrap the validation with conditions of network, e.g.
```
if navigator.onLine
  $auth.validateUser().then ->
    ...
else if localStorage['auth_headers']
  # offline with token, do not validate, and assume the token is still available
    ...
else
  # offline and no token
```

**Comment by dflourusso** [comment]
@Charlie-Hua Great solution, it can works, but supposed that navigator is online and happen some another network problem, like slow connection or DNS and the request finish by timeout, the token will be removed.

**Comment by scottward** [comment]
Good point @Charlie-Hua. That would cover quite a few cases, but not all as @dflourusso points at. Another case: if the server returns a 500, the user loses their credentials for when it's back up.

**Comment by Charlie-Hua** [comment]
@dflourusso Yes, you are right, some cases are not covered. Originally I was trying to recover auth_headers according to res.data from failed token auth request intercepted by angular. It's hacky and error prone.

But can we make sure that, `no res.data` = `no validation error`? 
How many cases need to be tested and how to test? Thank you!

**Comment by dflourusso** [comment]
You are right @Charlie-Hua 

I can not we make sure this. I'm using a more consistent condition now, if the `http_status` is `<= 0` i'm sure that there aren't a server response.

**Comment by booleanbetrayal** [comment]
Fixed via #339
